File: runtime/java/src/main/java/dev/argon/verilization/java_runtime/StandardCodecs.java
package dev.argon.verilization.java_runtime;

import java.io.IOException;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.ArrayList;
import java.util.Optional;

public abstract class StandardCodecs {
    private StandardCodecs() {}


    public static final Codec<BigInteger> natCodec = new Codec<BigInteger>() {

        @Override
        public BigInteger read(FormatReader reader) throws IOException {
            return VLQ.decodeVLQ(reader, false);
        }

        @Override
        public void write(FormatWriter writer, BigInteger value) throws IOException {
            VLQ.encodeVLQ(writer, false, value);
        }

    };

    public static final Codec<BigInteger> intCodec = new Codec<BigInteger>() {

        @Override
        public BigInteger read(FormatReader reader) throws IOException {
            return VLQ.decodeVLQ(reader, true);
        }

        @Override
        public void write(FormatWriter writer, BigInteger value) throws IOException {
            VLQ.encodeVLQ(writer, true, value);
        }

    };

    public static final Codec<Byte> i8Codec = new Codec<Byte>() {

        @Override
        public Byte read(FormatReader reader) throws IOException {
            return reader.readByte();
        }

        @Override
        public void write(FormatWriter writer, Byte value) throws IOException {
            writer.writeByte(value);
        }

    };

    public static final Codec<Short> i16Codec = new Codec<Short>() {

        @Override
        public Short read(FormatReader reader) throws IOException {
            return reader.readShort();
        }

        @Override
        public void write(FormatWriter writer, Short value) throws IOException {
            writer.writeShort(value);
        }

    };

    public static final Codec<Integer> i32Codec = new Codec<Integer>() {

        @Override
        public Integer read(FormatReader reader) throws IOException {
            return reader.readInt();
        }

        @Override
        public void write(FormatWriter writer, Integer value) throws IOException {
            writer.writeInt(value);
        }

    };

    public static final Codec<Long> i64Codec = new Codec<Long>() {

        @Override
        public Long read(FormatReader reader) throws IOException {
            return reader.readLong();
        }

        @Override
        public void write(FormatWriter writer, Long value) throws IOException {
            writer.writeLong(value);
        }

    };

    public static final Codec<String> stringCodec = new Codec<String>() {

        @Override
        public String read(FormatReader reader) throws IOException {
            BigInteger length = natCodec.read(reader);
            byte[] data = reader.readBytes(length.intValueExact());
            return new String(data, StandardCharsets.UTF_8);
        }

        @Override
        public void write(FormatWriter writer, String value) throws IOException {
            byte[] data = value.getBytes(StandardCharsets.UTF_8);
            natCodec.write(writer, BigInteger.valueOf(data.length));
            writer.writeBytes(data);
        }

    };

    public static <T> Codec<Optional<T>> optionalCodec(Codec<T> elementCodec) {
        return new Codec<Optional<T>>() {
            
            @Override
            public Optional<T> read(FormatReader reader) throws IOException {
                boolean present = reader.readByte() != 0;
                if(present) {
                    return Optional.of(elementCodec.read(reader));
                }
                else {
                    return Optional.empty();
                }
            }
    
            @Override
            public void write(FormatWriter writer, Optional<T> value) throws IOException {
                if(value.isPresent()) {
                    writer.writeByte((byte)1);
                    elementCodec.write(writer, value.get());
                }
                else {
                    writer.writeByte((byte)0);
                }
            }
    
        };
    }


}

File: runtime/scala/src/main/scala/dev/argon/verilization/scala_runtime/StandardCodecs.scala
package dev.argon.verilization.scala_runtime

import zio.{IO, ZIO, Chunk}
import java.nio.charset.StandardCharsets

object StandardCodecs {
    
    val natCodec: Codec[BigInt] = new Codec[BigInt] {
        override def read[R, E](reader: FormatReader[R, E]): ZIO[R, E, BigInt] =
            VLQ.decodeVLQ(reader, false)

        override def write[R, E](writer: FormatWriter[R, E], value: BigInt): ZIO[R, E, Unit] =
            VLQ.encodeVLQ(writer, false, value)
    }
    
    val intCodec: Codec[BigInt] = new Codec[BigInt] {
        override def read[R, E](reader: FormatReader[R, E]): ZIO[R, E, BigInt] =
            VLQ.decodeVLQ(reader, true)

        override def write[R, E](writer: FormatWriter[R, E], value: BigInt): ZIO[R, E, Unit] =
            VLQ.encodeVLQ(writer, true, value)
    }
    
    val i8Codec: Codec[Byte] = new Codec[Byte] {
        override def read[R, E](reader: FormatReader[R, E]): ZIO[R, E, Byte] =
            reader.readByte()

        override def write[R, E](writer: FormatWriter[R, E], value: Byte): ZIO[R, E, Unit] =
            writer.writeByte(value)
    }
    
    val i16Codec: Codec[Short] = new Codec[Short] {
        override def read[R, E](reader: FormatReader[R, E]): ZIO[R, E, Short] =
            reader.readShort()

        override def write[R, E](writer: FormatWriter[R, E], value: Short): ZIO[R, E, Unit] =
            writer.writeShort(value)
    }
    
    val i32Codec: Codec[Int] = new Codec[Int] {
        override def read[R, E](reader: FormatReader[R, E]): ZIO[R, E, Int] =
            reader.readInt()

        override def write[R, E](writer: FormatWriter[R, E], value: Int): ZIO[R, E, Unit] =
            writer.writeInt(value)
    }
    
    val i64Codec: Codec[Long] = new Codec[Long] {
        override def read[R, E](reader: FormatReader[R, E]): ZIO[R, E, Long] =
            reader.readLong()

        override def write[R, E](writer: FormatWriter[R, E], value: Long): ZIO[R, E, Unit] =
            writer.writeLong(value)
    }

    val stringCodec: Codec[String] = new Codec[String] {
        override def read[R, E](reader: FormatReader[R, E]): ZIO[R, E, String] =
            natCodec.read(reader).flatMap { length =>
                if(length > Int.MaxValue)
                    IO.die(new ArithmeticException("Length of string would overflow"))
                else
                    reader.readBytes(length.toInt)
            }.map { data => new String(data.toArray, StandardCharsets.UTF_8) }

        override def write[R, E](writer: FormatWriter[R, E], value: String): ZIO[R, E, Unit] = {
            val data = value.getBytes(StandardCharsets.UTF_8)
            natCodec.write(writer, value.length) *> writer.writeBytes(Chunk.fromArray(data))
        }
            
    }

    def listCodec[A](elementCodec: Codec[A]): Codec[Chunk[A]] = new Codec[Chunk[A]] {

        override def read[R, E](reader: FormatReader[R, E]): ZIO[R, E, Chunk[A]] =
            natCodec.read(reader).flatMap { length =>
                if(length > Int.MaxValue)
                    IO.succeed(length.toInt)
                else
                    IO.die(new ArithmeticException("Length of chunk would overflow"))
            }.flatMap { length =>
                def readElements(num: Int, data: Chunk[A]): ZIO[R, E, Chunk[A]] =
                    if(num > 0) elementCodec.read(reader).flatMap { elem => readElements(num - 1, data :+ elem) }
                    else IO.succeed(data)

                readElements(length, Chunk.empty)
            }

        override def write[R, E](writer: FormatWriter[R, E], value: Chunk[A]): ZIO[R, E, Unit] =
            natCodec.write(writer, value.size) *> ZIO.foreach_(value) { elem => elementCodec.write(writer, elem) }
            
    }

}
